counsel.el

```diff
@@ -1678,6 +1679,57 @@ When INITIAL-INPUT is non-nil, use it in the minibuffer during completion."
           (format "http://debbugs.gnu.org/cgi/bugreport.cgi?bug=%s"
                   (substring url 1)))))))
 
+(defvar counsel-url-expansions
+  nil
+  "List of (REGEXP . FORMAT) pairs.
+
+`counsel-url-expand' will expand the word at point according to
+FORMAT for the first matching REGEXP.  FORMAT can be either a
+string or a function.  If it is a string, it will be used as the
+format string for the `format' function, with the word at point
+as the next argument.  If it is a function, it will be called
+with the word at point as the sole argument.
+
+For example, a pair of the form:
+  '(\"^BSERV-[0-9]+$\" . \"https://jira.atlassian.com/browse/%s\")
+will expand to URL `https://jira.atlassian.com/browse/BSERV-100'
+when the word at point is BSERV-100.
+
+If the format element is a function, more powerful
+transformations are possible.  As an example,
+  '(\"^issue\\([0-9]+\\)$\" .
+    (lambda (word)
+      (concat \"http://debbugs.gnu.org/cgi/bugreport.cgi?bug=\"
+              (match-string 1 word))))
+trims the \"issue\" prefix from the word at point before creating the URL.")
+
+(defun compat-seq-some (pred lst)
+  "Call PRED on each element of LST and return first non-nil result.
+\(Implementation of seq-some for Emacs prior to version 25.)"
+  (catch 'result
+    (dolist (elt lst)
+      (let ((result (funcall pred elt)))
+        (when result
+          (throw 'result result))))
+    nil))
+
+(defun counsel-url-expand ()
+  "Expand word at point using `counsel-url-expansions'.
+The first pair in the list whose regexp matches the word at point
+will be expanded according to its format.  This function is
+intended to be used by `ivy-ffap-url-functions' to browse the
+result as a URL."
+  (let ((word-at-point (current-word)))
+    (compat-seq-some
+     (lambda (pair)
+       (let ((regexp (car pair))
+             (formatter (cdr pair)))
+         (when (string-match regexp word-at-point)
+           (if (functionp formatter)
```

Comment:
Just curious - any reason to prefer `functionp` over `fboundp` here?

Reply:
I'm not a frequent elisp writer, so I don't know the conventions. I preferred functionp because I know that formatter is bound here in the let, and I would tend to use fboundp when I don't know whether a symbol exists at all. But I recognize that we're talking about just the function slot of the symbol (or whatever the word is), so the distinction is hazy. If there's a reason or convention for preferring one over the other I'd be happy to follow it.

Reply:
Sorry, I should have looked into this more before asking. My habit is simply to ask the author first, because they may have already thought about it.

In [older Emacsen](https://ftp.gnu.org/old-gnu/Manuals/elisp-manual-20-2.5/html_chapter/elisp_12.html), `functionp` used to return `t` for any kind of function, including things which can't be passed to `funcall`, such as macros and special forms. Somehow I also managed to conflate `functionp` with doing "more than what is usually necessary" in my mind.

Contrast that with the [current description of `functionp`](https://www.gnu.org/software/emacs/manual/html_node/elisp/What-Is-a-Function.html) as returning `t` for any valid argument to `funcall`. In which case using `fboundp` instead of `functionp` here would simply be wrong.